docs: Add and refine developer contribution guidelines (#1356)
omegat-org/omegat: src_docs/developer/07.MannerOfThePullRequest.md

# Manner of the Pull Request

## Making Effective Comments on Pull Requests

Pull Requests (PRs) are not just about submitting code — they are also a space for discussion and collaboration. Writing clear, relevant, and respectful comments is essential to keep development productive and enjoyable for everyone.

This section is especially helpful for project managers, testers, and new contributors who want to provide feedback but are unsure how to do it effectively.

---

### 🎯 Stay Focused on the Scope

Each PR typically aims to fix a specific problem or introduce a clearly defined feature. When reviewing a PR, please:

- **Read the PR description** carefully. It explains the purpose and scope.
- **Comment only on things directly related** to that purpose.

> - If you discover an unrelated issue, note it down elsewhere (e.g., in a new ticket or in a chat), rather than posting it in the PR.

Author: Sometimes people add off-topic comments to a pull request because they get new ideas while reading it. These ideas might feel related to them, but they are not directly connected to the purpose of the pull request. To keep things clear and focused, it's better to discuss those ideas in a new issue or pull request.

```suggestion
- If you discover an unrelated issue, note it down elsewhere (e.g., in a new issue ticket or in a development mail list), rather than posting it in the PR.
- Even if an idea feels related because it came up while reading the PR, if it’s not directly part of the change being discussed, please keep it separate by opening a new issue or starting a new discussion.
```

> 📌 *Example: If the PR is fixing a spellchecker bug, avoid posting logs about an unrelated crash in the Machine Translation plugin.*

---

### 🧪 How to Report Testing Results

If you're testing a PR and something doesn't work as expected:

1. **First, confirm** whether the issue is part of the scope.
2. **Describe clearly** what you did, what happened, and what you expected.
3. **Avoid posting raw logs** unless they are directly related and help reproduce the problem.

> 🙋‍♀️ Instead of:  

> > *"Here’s a huge log dump I found — not sure if it's related."*
> >
> > ✅ Try this:  
> > *"When I tested the new caching feature with glossary lookup, I noticed that the refresh button didn’t seem to respond. Is that expected?"*
> 
> ---
> 
> ### 💬 Constructive Review Comments
> 
> When reviewing the code itself (even if you’re not a programmer), you can still help by:

Reviewer: ```suggestion
> *"Here's a huge log dump I found — not sure if it's related."*
>
> ✅ Try this:  
> *"When I tested the new caching feature with glossary lookup, I noticed that the refresh button didn't seem to respond. Is that expected?"*

---

### 💬 Constructive Review Comments

When reviewing the code itself (even if you're not a programmer), you can still help by:
```

- Asking for clarification when something is unclear.
- Noting typos, confusing wording, or behavior differences.
- Suggesting improvement *with context*, not just pointing out flaws.

> ❌ Avoid:  

> > "This doesn’t work."
> >
> > ✅ Prefer:  
> > "I expected the text to auto-save after editing, but it didn’t. Is this part of the feature or a known limitation?"

Reviewer: ```suggestion
> "This doesn't work."
>
> ✅ Prefer:  
> "I expected the text to auto-save after editing, but it didn't. Is this part of the feature or a known limitation?"
```

---

### 🤝 Respect Roles and Expertise

In small open-source teams, it's common for one person to wear many hats — a developer, tester, reviewer, and even
project manager. Still:

- **Let the PR author respond first** to comments or questions.

> - **Give the reviewer space** to evaluate the code’s quality and integration.

Reviewer: ```suggestion
- **Give the reviewer space** to evaluate the code's quality and integration.
```

- **Avoid interrupting the review** with unrelated findings.

This helps reduce confusion and avoids discouraging contributors who are generously offering their time and skill.

---

> 

Author: ```suggestion

### 🔄 Keep Development Moving — Don’t Block the Flow

Sometimes, reviewers suggest improvements like "please write better docs" or "can you update this part later?" These are helpful, but only if followed through. If the person making the suggestion doesn’t act — or if the contributor isn’t clear on what to do — progress stalls.

To keep development flowing:

- **Avoid vague promises.** If you suggest something, be ready to follow up or clearly hand off the task.
- **Support contributors.** If someone hesitates or is unsure, help them succeed — not stall.
- **Don’t “park” the PR** by saying "I'll write it later" and then forgetting.

> 🤝 A good reviewer helps motivate contributors and keeps energy moving forward. Be someone who builds momentum, not someone who blocks it.

---
```

Reviewer: > I want to add a section about a common but often unspoken issue - how vague promises or stalled feedback can block progress.

I agree, and I think this is very common when relying on volunteers. People do their best to help but that implies taking time away from everything else in their life with the main reward being that they've made themselves useful to strangers on the Internet.
It's a hard balance to find indeed: we need people to take part in the lifecycle of PRs but we should also respect their boundaries. That especially the case here where a lot of the volunteers are translators, and translators, like devs, are often freelancers, which means they can suddenly get an unexpected ton of work that will take them away from OmegaT

Maybe we could define a process to handle these situations instead?
Something like "if you leave a comment implying you will provide further feedback, you have a week (or another amount of time) to provide that feedback, then the author can send you a nudge and another week later, your comment can be marked as resolved"?

This will still block the PR for two weeks but will also provide a way forward without preventing people from getting involved in reviews.

### ✅ Summary

When commenting on PRs:

| DO ✅                             | DON'T ❌                            |
|----------------------------------|------------------------------------|
| Stay within the PR's scope       | Post unrelated issues or logs      |
| Write clear, respectful feedback | Make vague or accusatory remarks   |
| Report test results clearly      | Dump full logs without explanation |

> | Let the PR author reply          | Hijack the discussion              |

Author: ```suggestion
| Let the PR author reply          | Hijack the discussion              |
| Keep progress moving with clear actions   | Stall PRs with vague promises or silence |
```

> | Stay within the PR’s scope       | Post unrelated issues or logs      |
> | Write clear, respectful feedback | Make vague or accusatory remarks   |
> | Report test results clearly      | Dump full logs without explanation |
> | Let the PR author reply          | Hijack the discussion              |
> 
> We all want to improve OmegaT together — thoughtful communication helps us grow as a team and support each other’s contributions with mutual respect. 💙

Reviewer: ```suggestion
| Stay within the PR's scope       | Post unrelated issues or logs      |
| Write clear, respectful feedback | Make vague or accusatory remarks   |
| Report test results clearly      | Dump full logs without explanation |
| Let the PR author reply          | Hijack the discussion              |

We all want to improve OmegaT together — thoughtful communication helps us grow as a team and support each other's contributions with mutual respect. 💙
```

# Manner of the Pull Request

## Making Effective Comments on Pull Requests

> Pull Requests (PRs) are not just about submitting code — they are also a space for discussion and collaboration. Writing
> clear, relevant, and respectful comments is essential to keep development productive and enjoyable for everyone.

Reviewer: ```suggestion
Pull Requests (PRs) are not just about submitting code — they are also a space for discussion and collaboration. Writing clear, relevant, and respectful comments is essential to keep development productive and enjoyable for everyone.
```

> This section is especially helpful for project managers, testers, and new contributors who want to provide feedback but
> are unsure how to do it effectively.

Reviewer: ```suggestion
This section is especially helpful for project managers, testers, and new contributors who want to provide feedback but are unsure how to do it effectively.
```

---

### 🎯 Stay Focused on the Scope

Each PR typically aims to fix a specific problem or introduce a clearly defined feature. When reviewing a PR, please:

- **Read the PR description** carefully. It explains the purpose and scope.
- **Comment only on things directly related** to that purpose.

> - If you discover an unrelated issue, note it down elsewhere (e.g., in a new ticket or in a chat), rather than posting
>   it in the PR.

Reviewer: ```suggestion
- If you discover an unrelated issue, note it down elsewhere (e.g., in a new ticket or in a chat), rather than posting it in the PR.
```

> > 📌 *Example: If the PR is fixing a spellchecker bug, avoid posting logs about an unrelated crash in the Machine
> Translation plugin.*

Reviewer: ```suggestion
> 📌 *Example: If the PR is fixing a spellchecker bug, avoid posting logs about an unrelated crash in the Machine Translation plugin.*
```

---

### 🧪 How to Report Testing Results

If you're testing a PR and something doesn't work as expected:

1. **First, confirm** whether the issue is part of the scope.
2. **Describe clearly** what you did, what happened, and what you expected.
3. **Avoid posting raw logs** unless they are directly related and help reproduce the problem.

> 🙋‍♀️ Instead of:  
> *"Here’s a huge log dump I found — not sure if it's related."*
>
> ✅ Try this:  

> > *"When I tested the new caching feature with glossary lookup, I noticed that the refresh button didn’t seem to
> respond. Is that expected?"*

Reviewer: ```suggestion
> *"When I tested the new caching feature with glossary lookup, I noticed that the refresh button didn’t seem to respond. Is that expected?"*
```

> ❌ Avoid:  
> "This doesn’t work."
>
> ✅ Prefer:  
> "I expected the text to auto-save after editing, but it didn’t. Is this part of the feature or a known limitation?"

---

### 🤝 Respect Roles and Expertise

In small open-source teams, it's common for one person to wear many hats — a developer, tester, reviewer, and even
project manager. Still:

- **Let the PR author respond first** to comments or questions.
- **Give the reviewer space** to evaluate the code’s quality and integration.
- **Avoid interrupting the review** with unrelated findings.

This helps reduce confusion and avoids discouraging contributors who are generously offering their time and skill.

---

### ✅ Summary

When commenting on PRs:

| DO ✅                             | DON'T ❌                            |
|----------------------------------|------------------------------------|
| Stay within the PR’s scope       | Post unrelated issues or logs      |
| Write clear, respectful feedback | Make vague or accusatory remarks   |
| Report test results clearly      | Dump full logs without explanation |
| Let the PR author reply          | Hijack the discussion              |

> We all want to improve OmegaT together — thoughtful communication helps us grow as a team and support each other’s
> contributions with mutual respect. 💙

Reviewer: ```suggestion
We all want to improve OmegaT together — thoughtful communication helps us grow as a team and support each other’s contributions with mutual respect. 💙
```
